### ISSUE TITLE
TimerUITick

### ISSUE THREAD

**Issue by thebeerbaron2018** [nemosminer/NemosMiner#696]
Away from my rigs at the moment but I have been getting intermittent PS errors which cause an exception windows dialog and the PS script crashes totally.  It only seems to occur when the earnings tracker is enabled and I am trying to track down exactly when it happens.  I use only Zerg and Nicehash (no plus logic) and the error seems to be at line 336 with the exception relating to TimerUITick in NemosMiner.ps1 - although I suspect the error is actually in the TimerUITick code and not related to the Form Load code since the miner opens and beings mining correctly - the error appears only after a period of mining.  I'll update with more detail and screenshots of the error once I get back to the rigs.

**Comment by thebeerbaron2018** [comment]
Someone else posted the screenshot in the bitcointalk thread
![ip bitcointalk org](https://user-images.githubusercontent.com/36641730/42419929-aeb4dab8-82b5-11e8-83e8-b246a8b9ba94.png)

**Comment by MrPlusGH** [comment]
This one seems weird...
Failing to repro so far.

**Comment by thebeerbaron2018** [comment]
Yeah I am sure it's related to the earnings tracker and is intermittent.  Worst bugs to debug.  I'll run all my rigs on displays until one breaks down and try and track down more info for you.

**Comment by MrPlusGH** [comment]
Thanks for the help !

Never seen this one on my rigs and they've been running for days.

TimerUITick is a simple global function that refreshes the UI. 
No habla espanol :) but it looks like it's saying the function does not exists.

I'll monitor my rigs and try to catch it.

**Comment by thebeerbaron2018** [comment]
(again - not my screenshot) - I'll post an English shot of the error once I get home - the interesting thing is that the miner hasn't crashed on me in months of continuous operation.  Since the "new" balance section was introduced this exception in the white box has been happening - and as I said before - if you turn off the earningstracker by using false in the config - I think the problem goes away.  I did notice that when it crashed before the first error was something to do with the earnings part and then the timer error happened repeatedly afterwards - I'll need to catch the first error which is time consuming because I am not sure of the exact conditions that raise the error but I'll keep trying

**Comment by thebeerbaron2018** [comment]
OK so crashes even with the earnings tracker turned off.... I have an English screenshot and the exception text for you - and the corecycle log.... lots of errors for you there :) - the miner itself last console entry was "Loading pool stats...." - no errors in there at all and the same with the switching log.  Hope this helps

```
************** Exception Text **************
System.Management.Automation.CommandNotFoundException: The term 'TimerUITick' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at <ScriptBlock>(Closure , FunctionContext )
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseToInvoke clauseToInvoke, Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.<>c__DisplayClass57_0.<InvokeWithPipe>b__0()
   at System.Management.Automation.Runspaces.RunspaceBase.RunActionIfNoRunningPipelinesWithThreadCheck(Action action)
   at System.Management.Automation.ScriptBlock.InvokeWithPipe(Boolean useLocalScope, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Boolean propagateAllExceptionsToTop, List`1 variablesToDefine, Dictionary`2 functionsToDefine, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeAsDelegateHelper(Object dollarUnder, Object dollarThis, Object[] args)
   at System.Windows.Forms.Timer.OnTick(EventArgs e)
   at System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```
[CoreCycle.txt](https://github.com/nemosminer/NemosMiner/files/2173937/CoreCycle.txt)
![screenshot_2018-07-08_18-02-28](https://user-images.githubusercontent.com/36641730/42422190-b3ee11a6-82d9-11e8-8dad-8e48c7c4a48a.png)

**Comment by MrPlusGH** [comment]
[NPlusMiner.txt](https://github.com/nemosminer/NemosMiner/files/2173971/NPlusMiner.txt)

I still don't repro the issue but you seem to see it quite often.
Could you try the above NPlusMiner.ps1 version (rename from txt to ps1)

I simply changed from using a function to getting the script block straight in the timer.

@nemosminer , I'd suggest to rename this issue.

**Comment by thebeerbaron2018** [comment]
Sure I'll copy across to all my rigs and then the waiting starts.

Did I rename the issue title badly?

Should I change it something else?  What do you suggest?

Regards
Darren
________________________________

**Comment by thebeerbaron2018** [comment]
Pushed that file to half my rigs.  It complained about NMP.ICO which is named differently in nemo's release - but can ignore that.

Sadly they all get to "loading pool stats..." and hangs on all machines.


CORECYCLE:

**********************
Windows PowerShell transcript start
Start time: 20180708192925
Username: BORG2\miner
RunAs User: BORG2\miner
Configuration Name:
Machine: BORG2 (Microsoft Windows NT 10.0.17134.0)
Host Application: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -version 5.0 -executionpolicy bypass -windowstyle normal -command .\NemosMiner.ps1 -ConfigFile '.\Config\config.json'
Process ID: 7828
PSVersion: 5.1.17134.112
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.17134.112
BuildVersion: 10.0.17134.112
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************


MINER:

**********************
Windows PowerShell transcript start
Start time: 20180708192918
Username: BORG2\miner
RunAs User: BORG2\miner
Configuration Name:
Machine: BORG2 (Microsoft Windows NT 10.0.17134.0)
Host Application: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -version 5.0 -executionpolicy bypass -windowstyle normal -command .\NemosMiner.ps1 -ConfigFile '.\Config\config.json'
Process ID: 7828
PSVersion: 5.1.17134.112
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.17134.112
BuildVersion: 10.0.17134.112
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
Finding available TCP Port for NVIDIA
Miners API Port: 4068
Idle
File loaded - .\Core.ps1 -
File loaded - .\Include.ps1 -
File loaded - .\Include.ps1 - <ScriptBlock>
File loaded - .\Core.ps1 - <ScriptBlock>
Starting Cycle
Loading BTC rate from 'api.coinbase.com'..
Loading pool stats..

________________________________

**Comment by nemosminer** [comment]
been going through alot of changes over past few days you may not have complete source code, current source is tested and stable 

im testing it atm on multiple rigs no errors at all coming up2 24hours running now....

**Comment by thebeerbaron2018** [comment]
It's very odd - have been running Nemos since V3 with no issues at all.  It's only this last update to 3.3 that's caused this issue - have never seen the white backed .NET Windows error panel come up and my rigs run 24/7.  Obviously this is a problem for me to run unattended.  Since it's intermittent I am going to sit and watch all the rigs for 24 hours to find out the exact moment the error happens.  I have two possible theories : 1st is a memory leak (unlikely) and 2nd is I run a reboot script to reboot the rigs every 24hrs, I am wondering if the main panel is being closed but the script is still attempting to run which would explain all the errors relating to various fields on the form that would no longer exist.  I have tried rebooting manually however and that works without issue.  I am glad that the spanish guy noticed the same error as me - this indicates that it's not just my rigs that are noticing the issue though.

**Comment by nemosminer** [comment]
v3.3 is still work in progress... its not realeased

**Comment by thebeerbaron2018** [comment]
Yep.  I like to be cutting edge so am quite happy to clone the git / download the zip of the latest code (which is versioned 3.3).  I also don't mind assisting in fixing too so long as of use to you guys :)

**Comment by nemosminer** [comment]
cheers @thebeerbaron2018

**Comment by nemosminer** [comment]
i'll monitor my 6x1070ti test rigs over next 24-48hours see if i can reproduce

**Comment by MrPlusGH** [comment]
I have rigs with more than 3 days runtime and don't see the issue with NPlusMiner.
There must be something specific that we are yet to find.

Thanks a lot for the kindness and the help @thebeerbaron2018  ! We'll get this one!

**Comment by thebeerbaron2018** [comment]
Much-o coffee and sitting up all night for me.... it's obviously a combination of miner switch and core code that is causing it....  As you can see from the corecycle I posted above.... the script had been running for over 12 hours before it happened.  Got my monitoring hat on.

**Comment by thebeerbaron2018** [comment]
OK just watched one crash out me - first error that does not get reported is "Pipeline has been stopped".  Apparently this is due to slow running code - which makes sense because the "Loading pool stats..." seems to be MUCH slower on this code than it was before.  I am wondering whether something breaks if the script doesn't execute in a timely manner?  There was nothing obvious that this one miner was doing over my others?

Also, don't know if this is related (searching for pipeline and foreach-object said it could be a cause of pipeline breakage) but there are a lot of errors every cycle during the "comparing miners and pools" section:

```
Last Refresh: 07/08/2018 23:04:01
Sleeping 600
Starting Cycle
Loading BTC rate from 'api.coinbase.com'..
Loading pool stats..
Computing pool stats..
Loading miners..
Comparing miners and pools..
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
PS>TerminatingError(ForEach-Object): "You cannot call a method on a null-valued expression."
closing current miner and switching
Starting miner
Idle:  6
```
The recommended fix is 

```
#original code
Get-Something | Foreach-Object {do-something} 
# replace with this
foreach ($Obj in Get-Something) {
  do-something $Obj
}

```

**Comment by thebeerbaron2018** [comment]
So I think the problem may lie in lines 197-205 of core.ps1 - hope that narrows it down - am beginning to think it's definitely something to do with that block of code suddenly taking a lot longer (used to take 1-2 seconds - now takes 20+ seconds) for some reason.

**Comment by thebeerbaron2018** [comment]
Rigs all fell over with the same messages over night - now going to splat NPlus latest (4.1.0.1 from git clone zip) on half the machines and see if the problem persists there too.... already noticed that the "loading pools" is equally slow on the swapped machines - will let run to test for crashes

**Comment by MrPlusGH** [comment]
Please use the latest release for NPlusMiner. 
Master is constantly changing and not supported.

**Comment by thebeerbaron2018** [comment]
I know it's not supported but wanted to check if the same problem exists.  I can report that it does.  The only solution for me is to revert all code to the latest "release" build which is stable.  I'll keep a couple of rigs spare at your disposal if you want to test code with me further.

**Comment by MrPlusGH** [comment]
Sorry to hear this. I'll keep investigating.
Difficult though as I haven't seen any occurrence on my rigs.

**Comment by thebeerbaron2018** [comment]
It's definitely something to do with the new interface or the loading pool status part (although the code block for that section hasn't changed - but the variables block it draws information from has).

**Comment by MrPlusGH** [comment]
That's much of a guess. I've been able to generate other exceptions by trying to break code...
Could you try the version of NPlusMiner you'll find at the below link and let us know if it is any better?
https://github.com/MrPlusGH/NPlusMiner/archive/master.zip

**Comment by thebeerbaron2018** [comment]
OK downloaded and I am going to migrate to a 6 rig farm to test :)  Results later ....

**Comment by nemosminer** [comment]
I'll keep testing
Difficult though as I haven't seen any occurrence on my rigs.
testing updated UITick atm

**Comment by thebeerbaron2018** [comment]
Same error on two rigs so far..... it's all about the UI - the timer and other controls such as labels - it's almost like the form is disposed of and then the methods on those controls are then called and there isn't a form for it to use.  I have running another 6 machines with Nemo 3.2 on the old style UI and those are all absolutely rock stable 

```
Last Refresh: 07/09/2018 20:03:13
Sleeping 600
Starting Cycle
Loading BTC rate from 'api.coinbase.com'..
Loading pool stats..
The property 'Enabled' cannot be found on this object. Verify that the property exists and can be set.
At C:\Users\miner\Desktop\testing\NPlusMiner.ps1:108 char:5
+     $TimerUI.Enabled = $False
+     ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyNotFound
The property 'Enabled' cannot be found on this object. Verify that the property exists and can be set.
At C:\Users\miner\Desktop\testing\NPlusMiner.ps1:108 char:5
+     $TimerUI.Enabled = $False
+     ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyNotFound

You cannot call a method on a null-valued expression.
At C:\Users\miner\Desktop\testing\NPlusMiner.ps1:327 char:5
+     $TimerUI.Start()
+     ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
You cannot call a method on a null-valued expression.
At C:\Users\miner\Desktop\testing\NPlusMiner.ps1:327 char:5
+     $TimerUI.Start()
+     ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

The property 'Enabled' cannot be found on this object. Verify that the property exists and can be set.
At C:\Users\miner\Desktop\testing\NPlusMiner.ps1:108 char:5


```

**Comment by thebeerbaron2018** [comment]
Am trying something radical - I have changed the interval timer from 50 to 100 in the From_Load of NplusMiner to see if it's just a long loop thing that is causing it.... failing that - it might be best to do away with the time loop code as it's own interval and instead call on a recursive from the main loop - so if interval is set to 600 and you want to call every 50 seconds - run a pause 1 > add 1 to counter > pause 1 > add 1 to counter etc loop on the main thread and when counter hits 600 > refresh the form - that way it will be in sequence and not when the code is trying to collect pool information etc.... just a thought ...

**Comment by nemosminer** [comment]
try this: https://github.com/nemosminer/NemosMiner/archive/master.zip

**Comment by thebeerbaron2018** [comment]
Nothing out of the ordinary.  Asrock motherboard, 4gb ram, dual core celeron processor, WIN10 clean install, powershell 5.1.17134.112 - what concerns me is that the Spanish guy on bitcointalk was having the same problem so I know it's not just me - and that it's only started happening since the UI change - and that you may run into a lot of people finding it crashing on your next release build!  Uping the timer to 100 didn't solve the problem so what I am going to try now is stopping the timer altogether (and then stopping it between "Starting new cycle" and "sleeping") to see if that makes a difference.

**Comment by thebeerbaron2018** [comment]
By putting some extra logging lines in I am able to say that the crash (when it happens) always happens between 2 and 3 (ie, it shows 1 and 2 but never 3 in the log).  So the only thing that may be different here is that my location (as the guy from Spain) is EU and I imagine yours is US.... no idea why that should make a difference, but I have now set all rigs to US and retrying.... very weird issue that wasn't present in the old UI

**Comment by thebeerbaron2018** [comment]
OK changing to US didn't work either. Now looking more into the TimerUI scope.  Testing $Global:TimerUI because you reference the timer within functions and I am not sure whether the scope cascades down into the functions.

**Comment by nemosminer** [comment]
i have 3 different rigs testing latest source code atm

over 24hours running now will leave running for next few days to confirm stability

i have not been able to reproduce this at all 

(Code seems very stable) i have not had crashing or errors at all 100% stable

**Comment by MrPlusGH** [comment]
![image](https://user-images.githubusercontent.com/36303784/42558155-ac0af828-84f0-11e8-82ee-5f2e023fa938.png)

Still no occurrence of the issue.

**Comment by MrPlusGH** [comment]
NPlusMiner 4.2 fixes a performance issue with PoolsStats polling performances. Thanks for pointing this out!
I'm really not sure that there could be a link with the issue you're facing but it might be worth a try as you said it was crashing when updating pools.
It's currently flowing through autoupdate so if you have any version after 4.0, You can just start it and get the update.
https://github.com/MrPlusGH/NPlusMiner/releases

Let us know if this has any impact on your issue.
And thanks again for your help and involvement!

**Comment by thebeerbaron2018** [comment]
Cool!  Deploying to 6 rigs now - will let you know in a few hours

**Comment by thebeerbaron2018** [comment]
3 hours in and all rigs still up.  The "selecting pool" part is now the same as it was before and whatever you changed I think may have solved it.  Reserving judgement until the morning - if they're all still up by then then it's definitely good.

**Comment by MrPlusGH** [comment]
The main thing that has changed is that we had bad performances has the pool config variable is a script property that reads straight from disk. I changed code so we read it once and not 10 times for each pool.
Even if that was a real issue, I do not really see the link with the symptoms you were describing...

Fingers crossed :) We might be lucky on this one ;)

@nemosminer, you should get the latest version of BrainPlus and Pool files in NPlusMiner as even if it doesn't fix this particular issue (again, fingers crossed), it does really improve performances.

**Comment by thebeerbaron2018** [comment]
All rigs still running - for some reason that's fixed it!

**Comment by MrPlusGH** [comment]
WOW! so good to hear!
Please report back if it ever occur again.

Thanks a lot for the help you provided in debugging this one.
Always nice to have this sort of great help moreover on this type of issues which we can't repro and are kind of random.

**Comment by thebeerbaron2018** [comment]
I notice that during the "cycle" there is now a window refresh or redraw of the GUI.  I suspect this was the issue before - the code was referencing the controls on the form while it was being rebuilt.  I don't know whether this is due to me using low end CPU's (for power saving) on the rigs which are already under load from the validation on CPU of the GPU results - therefore the screen redraw is quite slow - I suspect that if the control referencing was happening during the redraw - if it didn't manage to complete the form before the code referenced the events or controls (which is why we saw multiple different errors like timerUItick and references to the status labels etc) then the PS script was aborting with the correct error that the controls didn't exist at that time.

**Comment by nemosminer** [comment]
https://github.com/nemosminer/NemosMiner/issues/732

**Comment by MrPlusGH** [comment]
https://github.com/MrPlusGH/NPlusMiner/issues/110

**Comment by Olegik-ah** [comment]
Same, but little different error...

************** Exception Text **************
System.Management.Automation.RuntimeException: The property 'Lines' cannot be found on this object. Verify that the property exists and can be set.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at <ScriptBlock>(Closure , FunctionContext )
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseToInvoke clauseToInvoke, Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.<>c__DisplayClass57_0.<InvokeWithPipe>b__0()
   at System.Management.Automation.Runspaces.RunspaceBase.RunActionIfNoRunningPipelinesWithThreadCheck(Action action)
   at System.Management.Automation.ScriptBlock.InvokeWithPipe(Boolean useLocalScope, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Boolean propagateAllExceptionsToTop, List`1 variablesToDefine, Dictionary`2 functionsToDefine, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeAsDelegateHelper(Object dollarUnder, Object dollarThis, Object[] args)
   at System.Windows.Forms.Timer.OnTick(EventArgs e)
   at System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

**Comment by thebeerbaron2018** [comment]
@MrPlusGH sorry for the necro - just wanted to say whatever changes you have made have now definitely corrected this issue.  Have had 6 rigs running for 48 hours without fail so I am now in the process of deploying to all my rigs again.  Thank you for your hard work and am happy to be back again

**Comment by irimiab** [comment]
Hello

Yesterday I updated my Windows 10 Pro with the latest updates and I also updated the NVidia drivers to the latest version and NemosMiner to the latest version. I was using version 3.3 before.

Since the update, I keep getting this error every 3-4 hours or so. I tried version 3.3 and it does the same. So it's not a bug in NemosMiner, but something related with PowerShell, .NET Framework and Windows itself. I understand it has something to do with the performance of the system; I am running a system with 6 GeForce GTX 1070 video cards, but with a Celeron processor (Intel(R) Celeron(R) CPU G3930 @ 2.90GHz). The UI of NemosMiner indeed is a little slow.

What can I do? (except upgrading the processor 😬 )

**Comment by MrPlusGH** [comment]
@irimiab 
Could you try NPlusMiner and report if you see the same issue?
   https://github.com/MrPlusGH/NPlusMiner
Users have reported to not have the issue with NPlusMiner. Nemo and I are working on finding out why.
If you want to avoid benchmark, you can copy the stats folder over.

+MrPlus

**Comment by irimiab** [comment]
Yes, I will try. Currently I'm running MPM, which doesn't have this issue (obviously, since it's related with the UI).

**Comment by irimiab** [comment]
Tested the latest version of NPlusMiner (the today's release) and does exactly the same as NemosMiner.
Will have to migrate back to MPM.
